src/wallet/rpcwallet.cpp

```diff
@@ -161,8 +161,8 @@ UniValue getnewaddress(const JSONRPCRequest& request)
 
     // Generate a new key that is added to wallet
     CPubKey newKey;
-    if (!pwallet->GetKeyFromPool(newKey)) {
-        throw JSONRPCError(RPC_WALLET_KEYPOOL_RAN_OUT, "Error: Keypool ran out, please call keypoolrefill first");
+    if (!pwallet->GetKeyFromPool(newKey) || (pwallet->IsHDEnabled() && !pwallet->HasUnusedKeys(GetArg("-keypoolcritical", DEFAULT_KEYPOOL_CRITICAL)))) {
```

Comment:
In commit "[wallet] return error from getnewaddress if keypool drops to critical"

Should keypool size be checked before calling GetKeyFromPool instead of after, to avoid getting a key and throwing it away?